### PR TITLE
Adding an admin panel for the labels

### LINKED ISSUE #812: Re-integrate Labels for Exercises
Labels should be used to tag exercises with various information (such as a concept). Previously, CodeHarbor already had support for labels, so that we should re-enable them:

- [x] Allow users to add (existing) labels to an exercise
- [x] Show labels when viewing an exercise
- [x] Integrate labels into the search (so that one can filter based on labels)

In this first step, I would not go much further and build a UI for creating labels. Those labels should either be created automatically or through admins or something else to be defined.

### PR DESCRIPTION
**Pull request by Mathis-Z**
This adds an admin panel to manage the labels.
The admin panel and the select2 ajax requests now use the same LabelsController#search endpoint although the select2 does not need the additional information the admin panel uses. Maybe those should be separated into two endpoints?

For the tests I used feature specs because the admin panel has a lot of coffeescript code that I wanted to cover as well but I am not sure if its worth the increased test runtime. I tried to run the expensive "visit_page - clicking - form filling etc." logic just once per context to save some time but using "before(:all)" did not work with Capybara. The easiest solution would probably be to just merge all examples so there is only one example per context. What do you think?

This should also close #812

### CODE REVIEW

File: spec/features/labels_controller_spec.rb [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        end
+        wait_for_ajax
+      end
+
+      let!(:labels) { create_list(:label, 5) }
+      let!(:task) { create(:task, labels: labels[1..3]) }
+
+      it 'deletes the label' do
+        deletion_action
+        expect(Label.find_by(id: labels[1].id)).to be_nil
```
Comment (MrSerth) [inline spec/features/labels_controller_spec.rb:102]: I know you switched to the `find_by(id: ...)` syntax to avoid a `RecordNotFound` exception. But you could also switch to `where` and check with `.not_to exist`.

File: app/assets/javascripts/labels_index.coffee [hunk 2]
```diff
@@ -0,0 +1,183 @@
+class LabelsTable
```
Comment (MrSerth) [inline app/assets/javascripts/labels_index.coffee:1]: I like that you opted for a class here and therefore encapsulated a lot of code - thank you! At first, I was a bit surprised by the fact that you did everything on your own (including the sorting), but I understand that we don't have a nice alternative at the moment. Zuhra is currently working on upgrading bootstrap, which will also allow us to integrate Webpack (by using Shakapacker) and therefore any NPM package. For example, the [`sorttable` package](https://www.npmjs.com/package/sorttable) is the one that we are using in CodeOcean, even though it is already six years old.

Nevertheless, I also considered whether and how this large amount of CoffeeScript could have been reduced / prevented. And actually, since you also extended the sorting to query the server-side objects dynamically, a package such as `sorttable` wouldn't help much (this specific package will well reorder the list, but this state won't be used directly to query new data). 

For modern JavaScript apps, there would be a data binding that would fill the table based on the given data, but this a not yet integrated in CodeHarbor and also wouldn't save too much. The Rails way (as a backend framework) would rather be to render the table using a Slim partial and just swapping the final HTML (probably, this would have been the way I'd choose, but just because I am not a frontend engineer).

Long story short: There are many thoughts I had on this file, but conclusively I think it is fine to leave it as it is. Maybe just take my main concern with you, that we would normally try to leverage the power of libraries and frameworks for those issues that have been solved already many times.

**Review by MrSerth — COMMENTED** [review]
Nice, thank you! This looks and works fine, just as expected. Don't be afraid, I still added a few comments, but nothing too big, I hope.

> The admin panel and the select2 ajax requests now use the same LabelsController#search endpoint although the select2 does not need the additional information the admin panel uses. Maybe those should be separated into two endpoints?

Generally speaking, I think it is fine to have them combined into one. Still, I made a few suggestions how we could further improve the `#search` endpoint. Since I suggested removing the information that is not necessary for a regular user, we could also consider sending an additional parameter to the search endpoint and validating that only admins (or others being able to access this new page) are allowed to use it.

> The easiest solution would probably be to just merge all examples so there is only one example per context. What do you think?

While I am well aware of the additional time, I think that this time should be compared to the whole developer experience, which is not limited to the execution time. There are also other factors involved, such as readability (of the tests) and descriptive error messages in case a test doesn't work as expected. That having said, I actually won't merge those, but rather leave them in the current status. It is rather unexpected for someone to execute all tests in the project, so that developers can mostly ignore the longer execution time (when working on something unrelated).

File: app/controllers/labels_controller.rb [hunk 3]
```diff
@@ -2,12 +2,54 @@
...
+
+  def index
+    @search = {}
+  end
+
+  def merge
+    label_ids = params[:label_ids]
+    new_name = params[:new_label_name]
+
+    if new_name.length.positive? && new_name.length < 15
```
[the remark this answers is not in the document]
Comment (Mathis-Z) [reply app/controllers/labels_controller.rb:18]: Since this is also being verified in the frontend, should we set the maximum label length as a constant in the label model and fill it into the frontend js page using embedded ruby?

[the remark this answers is not in the document]
Comment (MrSerth) [reply app/controllers/labels_controller.rb:18]: This would work and could be injected in the CoffeeScript file (remember to change the extension, e.g., make it `.js.coffee.erb`). As an alternative, you could also pass the number through a `data` attribute on an HTML node.

File: app/controllers/labels_controller.rb [hunk 4]
```diff
@@ -2,12 +2,54 @@
...
+      new_label = Label.create(name: new_name)
+      new_label.tasks = Task.where(id: TaskLabel.select(:task_id).where(label_id: label_ids).distinct)
+      Label.where(id: label_ids).destroy_all
+    else
+      head :bad_request
```
[the remark this answers is not in the document]
Comment (Mathis-Z) [reply app/controllers/labels_controller.rb]: #merge now returns :unprocessable_entity on all errors. If you look into the function, you will see that I changed how the transaction rollback is handled again. After reading some threads on StackOverflow I think this is how you are supposed to handle a transaction rollback in rails.

File: config/locales/en.yml [hunk 5]
```diff
@@ -820,12 +820,21 @@ en:
...
     new:
       header: "New Label"
       selection_suffix: " (new)"
+    index:
+      merge: 'Merge And Rename To'
+      delete: 'Delete selected'
+      change_color_to: 'Change Color To'
+      delete_confirmation: 'This will delete the {selected_labels_count} selected label(s).'
+      merge_confirmation: 'This will merge the {selected_labels_count} selected label(s) and rename them to "{new_merged_name}"'
+      change_color_confirmation: 'This will change the color of the {selected_labels_count} selected label(s).'
```
[the remark this answers is not in the document]
Comment (Mathis-Z) [reply config/locales/en.yml]: I did not know that existed. Thanks for the suggestion!

File: config/routes.rb [hunk 6]
```diff
@@ -60,8 +60,14 @@
...
   end
 
-  resources :labels, only: [] do
-    get :search, on: :collection
+  resources :labels, only: %i[index] do
+    collection do
+      get :search
+      get :list
+      get :merge
+      get :set_color
```
[the remark this answers is not in the document]
Comment (Mathis-Z) [reply config/routes.rb]: I replaced the set_color action with the regular update action like you suggested.

File: app/controllers/labels_controller.rb [hunk 7]
```diff
@@ -2,12 +2,54 @@
...
+
+  def search # rubocop:disable Metrics/AbcSize
+    paginated = Label.ransack(params[:search]).result.paginate(per_page: 3, page: params[:page])
 
     render json: {
-      results: paginated.map {|l| {id: l.name, text: l.name, label_color: l.color, label_font_color: l.font_color} },
+      results: paginated.map do |l|
+        l.attributes.merge(
+          font_color: l.font_color,
+          used_by_tasks: l.tasks.count,
```
[the remark this answers is not in the document]
Comment (Mathis-Z) [reply app/controllers/labels_controller.rb]: I changed the code to preload the tasks and only return the tasks.size information for admins that request it. The other fields are returned for all users and requests though.

File: app/controllers/labels_controller.rb [hunk 8]
```diff
@@ -2,12 +2,54 @@
...
-  def search
-    results = Label.ransack(name_i_cont: params[:search]).result
-    paginated = results.paginate(per_page: 3, page: params[:page])
+
+  rescue_from CanCan::AccessDenied do |_exception|
+    redirect_to root_path, alert: t('controllers.authorization')
+  end
+
+  def index
+    @search = {}
```
[the remark this answers is not in the document]
Comment (Mathis-Z) [reply app/controllers/labels_controller.rb]: That is another leftover of the list endpoint that I dropped along the way. Again, sorry for not checking the final code thoroughly enough.

File: spec/controllers/labels_controller_spec.rb [hunk 9]
```diff
@@ -27,14 +27,21 @@
         expect(JSON.parse(response.body, symbolize_names: true)).to eql(
           {
             pagination: {more: false},
-            results: [{id: label.name, label_color: label.color, label_font_color: label.font_color, text: label.name}],
+            results: [
+              label.attributes.merge(
+                font_color: label.font_color,
+                used_by_tasks: label.tasks.count,
+                created_at: label.created_at.to_fs(:rfc822), # LabelsController#search changes the time format to rfc822 before returning the labels
+                updated_at: label.updated_at.to_fs(:rfc822)
```
[the remark this answers is not in the document]
Comment (Mathis-Z) [reply spec/controllers/labels_controller_spec.rb]: I moved it into the model but did not write a model spec for it. It is tested indirectly in the labels_controller_spec though. Do you want me to write another model test for that?

File: spec/features/labels_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+      let!(:matching_labels) { [create(:label, name: 'Loops'), create(:label, name: 'Java loops')] }
+      let!(:not_matching_labels) { [create(:label, name: 'label1'), create(:label, name: 'python for loop')] }
+
+      it 'filters by name correctly' do
+        visit(labels_path)
+        fill_in id: 'label-name-filter-input', with: 'loops'
+        find('input', id: 'label-name-filter-input').send_keys(:enter)
+        wait_for_ajax
+        matching_labels.each {|label| expect(page).to have_selector('.labels-table > tbody > tr > td > .task_label', text: label.name) }
+        not_matching_labels.each {|label| expect(page).not_to have_selector('.labels-table > tbody > tr > td > .task_label', text: label.name) }
```
[the remark this answers is not in the document]
Comment (Mathis-Z) [reply spec/features/labels_controller_spec.rb]: I though being overspecific was better than being not specific enough to avoid false negatives in testing. I shortened the css a bit to make it more readable.

File: spec/features/labels_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+          accept_prompt do
+            click_button('merge-labels-button')
+          end
+          wait_for_ajax
+        end
+
+        it 'renames the label' do
+          renaming_action
+          expect(Label.find_by(name: new_name)).not_to be_nil
+        end
```
[the remark this answers is not in the document]
Comment (Mathis-Z) [reply spec/features/labels_controller_spec.rb]: Since the merge action changed in the new commit, this test example has now changed too. It now checks for a specific label if the name has changed so the test works even if there is another label with the same name before the renaming_action. This is still a case that I did not consider though. What should happen if after a renaming_action two labels have the same name? Is that acceptable and the responsibility of the admins or should this be prevented? My intuition would be to let the admins handle this. What do you think?

File: app/assets/javascripts/labels_index.coffee (hunk 2 above)
Comment (Mathis-Z) [reply app/assets/javascripts/labels_index.coffee:1]: I do not think there is a library used in CodeHarbor that could have been used for this and I did not want to use a new one for what I though would be easily solved with some coffeescript. It did end up more complex than I expected though and I will certainly try to use a library/framework next time. Thank you for your feedback!

File: app/models/label.rb [hunk 12]
```diff
@@ -59,7 +60,23 @@ def choose_label_color
...
+  def to_h
+    {
+      id:,
+      name:,
+      color:,
+      font_color:,
+      used_by_tasks: tasks.loaded? ? tasks.size : 0,
+      created_at: created_at.to_fs(:rfc822),
+      updated_at: updated_at.to_fs(:rfc822),
+    }
```
Comment (Mathis-Z) [inline app/models/label.rb:72]: another way would be to use
```
{
...
used_by_tasks: (tasks.size if tasks.loaded?),
...
}.compact
```
This way the used_by_tasks key would not be in there at all which is more explicit than just silently setting used_by_tasks to 0. What do you think?

File: app/controllers/labels_controller.rb [hunk 13]
```diff
@@ -2,13 +2,43 @@
...
+
+  def index; end
+
+  def update
+    @label.update(params.permit(:color))
+  end
+
+  def destroy
+    @label.destroy
+  end
```
Comment (Mathis-Z) [inline app/controllers/labels_controller.rb:18]: Should we use update! and destroy! instead, catch the error and return the error message?

File: app/controllers/labels_controller.rb [hunk 14]
```diff
@@ -2,12 +2,54 @@
...
+      new_label.tasks = Task.where(id: TaskLabel.select(:task_id).where(label_id: label_ids).distinct)
+      Label.where(id: label_ids).destroy_all
+    else
+      head :bad_request
+    end
+  end
+
+  def set_color
+    new_color = params[:new_color]
+    if /^[0-9a-fA-F]{6}$/.match?(new_color)
```
[the remark this answers is not in the document]
Comment (MrSerth) [reply app/controllers/labels_controller.rb]: Usually, just failing silently in the backend is not recommended. What if we update the controller as follows:

```rb
    if @label.update(params.permit(:color))
      render json: @label.to_h
    else
      render json: @label.errors, status: :unprocessable_entity
    end
```

File: app/assets/javascripts/labels_index.coffee [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+
+    if (confirm(confirm_msg))
+      $.ajax({
+        url: "/labels/merge",
+        type: "post",
+        data: {
+          label_ids: @selected_label_ids,
+          new_label_name: new_label_name
+        },
+        success: (response) => @reload_table();
```
Comment (MrSerth) [inline app/assets/javascripts/labels_index.coffee:152]: What about handling an error case in the frontend as well? An admin won't receive any information if updating doesn't work as expected.... (I think an alert would be enough).

File: app/assets/javascripts/labels_index.coffee [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+      requests = []
+      for id in @selected_label_ids
+        requests.push(
+          $.ajax({
+            url: "/labels/#{id}",
+            type: "patch",
+            data: { color: new_color }
+          })
+        );
+      $.when.apply($, requests).then( () => @reload_table() );
```
Comment (MrSerth) [inline app/assets/javascripts/labels_index.coffee:173]: A bit more complicated here, sure, but some error handling might be nice 😉

File: spec/controllers/labels_controller_spec.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (MrSerth) [reply spec/controllers/labels_controller_spec.rb]: I think, that's fine; the controller test covers these lines as well.

File: spec/features/labels_controller_spec.rb (hunk 11 above)
[the remark this answers is not in the document]
Comment (MrSerth) [reply spec/features/labels_controller_spec.rb]: I would actually argue that Rails can handle this for us 😁. Since we are assuming that the name is unique (for example, when assigning labels), we should also ensure that they are unique.

What about adding the following to the name validation in the `label.rb` model?

```rb
, uniqueness: {case_sensitive: false}
```

Then, we don't need to consider something manually, but still have the desired result. 
<hr>

Otherwise, the new test looks fine and removes the issue I saw before 👍

File: app/assets/javascripts/labels_index.coffee (hunk 2 above)
Comment (MrSerth) [reply app/assets/javascripts/labels_index.coffee:1]: You're welcome, thanks for considering this :)
While a library might have helped here, it also comes at a cost: Just adding a library for every code fragment we need to write is also not an optimal solution -- we need to consider the expected maintenance costs and the increased dependency to something external over the additional complexity involved when writing it our own. So, this one comes with advantages and disadvantages as well...

File: app/models/label.rb (hunk 12 above)
Comment (MrSerth) [reply app/models/label.rb:72]: This sounds even better, actually. I also thought about returning `nil` rather than `0`, but omitting the key is even smarter.

File: config/locales/en.yml [hunk 17]
```diff
@@ -820,12 +820,21 @@ en:
       back: "Back"
       #submit:
     can_only_create_5: 'You can only add 5 labels'
+    too_long: 'The entered label name is too long'
+    cannot_be_empty: 'The entered label name is empty'
```
Comment (MrSerth) [inline config/locales/en.yml:824]:
```suggestion
    too_long: 'The label name entered is too long'
    cannot_be_empty: 'The label name entered is empty'
```

File: config/locales/en.yml [hunk 18]
```diff
@@ -820,12 +820,21 @@ en:
...
 
     edit:
       header: "Editing Label"
     new:
       header: "New Label"
       selection_suffix: " (new)"
+    index:
+      merge: 'Merge And Rename To'
+      delete: 'Delete selected'
+      change_color_to: 'Change Color To'
```
Comment (MrSerth) [inline config/locales/en.yml:834]: I would go with [Capitalization](https://capitalizemytitle.com/), similar to a paper here:

```suggestion
      merge: 'Merge and Rename to'
      delete: 'Delete selected'
      change_color_to: 'Change Color to'
```

File: spec/controllers/labels_controller_spec.rb [hunk 19]
```diff
@@ -19,22 +22,109 @@
...
+        expect(response).to redirect_to(root_url)
+      end
+    end
+
+    context 'when signed in as admin' do
+      let(:admin) { create(:admin) }
+
+      before { sign_in admin }
+
+      context 'with invalid new label name' do
```
Comment (MrSerth) [inline spec/controllers/labels_controller_spec.rb:112]:
```suggestion
      context 'with an invalid new label name' do
```

File: spec/controllers/labels_controller_spec.rb [hunk 20]
```diff
@@ -19,22 +22,109 @@
       end
     end
 
-    context 'with valid params' do
-      before { sign_in user }
-
-      it 'returns valid json' do
+    shared_examples 'json without tasks count' do
```
Comment (MrSerth) [inline spec/controllers/labels_controller_spec.rb:25]: Of course, this also applies to the `include_examples`:

```suggestion
    shared_examples 'json without the number of referrencing tasks' do
```

File: spec/features/labels_controller_spec.rb [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+describe 'LabelsController', js: true do
+  let(:admin) { create(:admin) }
+  let(:password) { attributes_for(:admin)[:password] }
+
+  before do
+    sign_in_with_js_driver(admin, password)
+  end
+
+  describe '#index' do
+    context 'when using name filter' do
```
Comment (MrSerth) [inline spec/features/labels_controller_spec.rb:14]:
```suggestion
    context 'when using the name filter' do
```

**Review by MrSerth — COMMENTED** [review]
Thanks for incorporating my feedback; I really like the cleaner implementation! 💪 
While I don't have much to share besides the various comments already submitted, here are a few wording suggestions. I am not a native English speaker either, so take them with a grain of salt 🧂

File: app/controllers/labels_controller.rb (hunk 14 above)
[the remark this answers is not in the document]
Comment (Mathis-Z) [reply app/controllers/labels_controller.rb]: That is indeed better than what I suggested a few threads below (using update! and rescuing the error).

File: spec/features/labels_controller_spec.rb (hunk 11 above)
[the remark this answers is not in the document]
Comment (Mathis-Z) [reply spec/features/labels_controller_spec.rb]: I will add a test for that too.

File: app/controllers/labels_controller.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (Mathis-Z) [reply app/controllers/labels_controller.rb]: There is also RecordNotFound that can occur if non-existing label_ids are submitted. Unfortunately ActiveRecord::RecordNotFound has no .record.errors.full_messages so I would suggest to just use e.message.
With RecordNotFound there are now 4 errors that we catch. I think we could safely catch the parent ActiveRecord::ActiveRecordError instead.

File: spec/features/labels_controller_spec.rb (hunk 11 above)
[the remark this answers is not in the document]
Comment (Mathis-Z) [reply spec/features/labels_controller_spec.rb]: I just noticed there is a case now where merging labels will fail even though it is supposed to work. If you select two labels a,b and try to merge them with new name 'b' the merge will fail. This is because #merge first tries to rename a to b but this fails because b already exists.
(If you select b,a and rename to b it works as expected)
One way to solve this would be to switch the order of destroying the old labels and updating the merged label:
```
ApplicationRecord.transaction do
  tasks = Task.where(id: TaskLabel.select(:task_id).where(label_id: label_ids))
  Label.destroy(label_ids[1..])
  Label.find(label_ids.first).update!(name: params[:new_label_name], tasks:)
rescue ActiveRecord::ActiveRecordError => e
  render json: e.message, status: :unprocessable_entity
  raise ActiveRecord::Rollback
end
```

File: app/controllers/labels_controller.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (MrSerth) [reply app/controllers/labels_controller.rb]: Yeah, I was checking the error classes as well and also identified that `RecordNotFound` doesn't provide `e.record` (which makes sense, since there is no record). I was also considering `ActiveRecordError`, but thought it would be too broad (also covering errors such as `ConnectionNotEstablished` or `StatementInvalid`), which we probably don't want to catch. Further, the `RecordNotFound` error should already be handled globally, simply returning a 404 (but of course, this can be validated again).

My other idea was to validate the potential changes (especially the update) or just add a second `rescue` block for other errors, such as the `RecordNotFound`.

File: spec/features/labels_controller_spec.rb (hunk 11 above)
[the remark this answers is not in the document]
Comment (MrSerth) [reply spec/features/labels_controller_spec.rb]: Ah, that's a good finding. Yes, changing the order sounds reasonable (keep in mind that the transaction guards us against potential errors, so either the merging completes successfully or nothing gets changed).

File: app/controllers/labels_controller.rb [hunk 22]
```diff
@@ -10,24 +10,32 @@ class LabelsController < ApplicationController
...
-        tasks: Task.where(id: TaskLabel.select(:task_id).where(label_id: label_ids))
-      )
+      tasks = Task.where(id: TaskLabel.select(:task_id).where(label_id: label_ids))
       Label.destroy(label_ids[1..])
-    rescue StandardError => e
-      render json: e, status: :unprocessable_entity
+      Label.find(label_ids.first).update!(name: params[:new_label_name], tasks:)
+    rescue ActiveRecord::RecordNotFound => e
+      flash.now[:alert] = e.message
+      render json: e.message, status: :unprocessable_entity
```
Comment (MrSerth) [inline app/controllers/labels_controller.rb:34]: If the record is not found, we might also return a 404, or specifically a `:not_found` -- or would you stay with the 422 (which would be fine as well)?

**Review by MrSerth — APPROVED** [review]

File: app/controllers/labels_controller.rb (hunk 22 above)
Comment (Mathis-Z) [reply app/controllers/labels_controller.rb:34]: I think it is okay to leave it as is. When one or more of the labels are not found the merge query is not processable as a whole.

Comment (MrSerth) [reply app/controllers/labels_controller.rb:34]: Alright, fine for me. Thanks!
